File: easy/arrays/cpp/palindrome_number_9.cpp
#include <stdio.h>
#include <stdbool.h>
#include <iostream>

class Solution {
	public: 
		Solution();

		bool isPalindrome(int x);
		~Solution();
};

Solution::Solution(){};
Solution::~Solution(void){}

bool Solution::isPalindrome(int x)
{
	if (x < 0)
	{
		return false;
	}

	int digits = 0;
	int reverse = 0;
	int original = x;

	// reverse the number here 
	while(x > 0)
	{
		digits = x % 10;

		reverse = reverse * 10 + digits;

		x = x/10;
	}
	return original == reverse;
}

int main(void)
{

	Solution palindrome;
	
	bool status = palindrome.isPalindrome(121);
	printf("%d\n", status);

	return 0;
}

File: easy/arrays/rust/palindrome_number_9.rs
pub struct Solution; 
impl Solution {
    pub fn is_palindrome(x: i32) -> bool {

        if x < 0 {
            return false;
        }
       let mut reverse: i32 = 0;
       let mut original: i32 = x;

       while original > 0{
       
           let digits = original % 10;
           
/*
 // check for potential integer overflow
           if reverse > i32::MAX / 10 || (reverse == i32::MAX / 10 && digits > 7){
                return false;
           }
*/
           reverse = reverse * 10 + digits;


        
           original/=10;
       }

       x == reverse 

    }
}

fn main()
{
    let status = Solution::is_palindrome(121);

    println!("Solution: {}", status);


}
